### ISSUE TITLE
acme/lego causes account rate limits 

### ISSUE THREAD

**Issue by rkoe** [NixOS/nixpkgs#85861]
**Describe the bug**
When upgrading to 20.03, the change from simp_le to lego causes
- all certificates to be re-created (which is not nice, since the old certificates would still work)
- if more than 10 certificates are used on a server, some certificates cannot re-created, because lego creates a new account per certificate an so runs into the account rate limit (max. 10 accounts in 3h per IP) -- and so the site-encryption breaks for servers with > 10 certificates.

According to #85152 this should be fixed -- but it isn't.

**To Reproduce**
1. Use a server with e.g. NixOS 19.09 and > 10 certificates.
2. Upgrade to 20.03
3. The upgrade will result in failing acme-services.

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.114, NixOS, 20.03.1445.95b9c99f6d0 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - channels(root): `"nixos-20.03.1445.95b9c99f6d0"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by arianvp** [comment]
Hmmm Maybe it's actually recreating the account every time... let me debug and see if that's the case...

**Comment by arianvp** [comment]
Doesn't seem to be recreating the account. So it indeed seems youre running into `certificate` ratelimits instead of account rate limits. Would you be so kind to provide logs as well?


```
[root@arianvp:/var/lib/acme/.lego]# sha256sum accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/account.json 
b40a96eecd5c6750ea8a3ec59ae8f1b5387057ae3bd84c82cbc01ec0e5a97850  accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/account.json
[root@arianvp:/var/lib/acme/.lego]# sha256sum accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/keys/arian.vanputten@gmail.com.key 
4ce788e135006e243ebf6af8c2f40c44fa3dad4b09deda45abdbbdba8e017701  accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/keys/arian.vanputten@gmail.com.key


[root@arianvp:/var/lib/acme/.lego]# mv techstock.photos techstock.photos.bak.bak

[root@arianvp:/var/lib/acme/.lego]# systemctl restart acme-techstock.photos.service 

[root@arianvp:/var/lib/acme/.lego]# sha256sum accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/account.json 
b40a96eecd5c6750ea8a3ec59ae8f1b5387057ae3bd84c82cbc01ec0e5a97850  accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/account.json
[root@arianvp:/var/lib/acme/.lego]# sha256sum accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/keys/arian.vanputten@gmail.com.key 
4ce788e135006e243ebf6af8c2f40c44fa3dad4b09deda45abdbbdba8e017701  accounts/acme-v02.api.letsencrypt.org/arian.vanputten@gmail.com/keys/arian.vanputten@gmail.com.key


```

**Comment by arianvp** [comment]
`extraDomains` does have the issue at the moment that once you set it; it isn't updated anymore. https://github.com/NixOS/nixpkgs/issues/84633 (see last bullet point)

The fix for this I will probably try to get backported

**Comment by rkoe** [comment]
> So it indeed seems youre running into `certificate` ratelimits instead of account rate limits.

No, I'm running into accout rate limits (I've replaced the real domain name with example.com):

```
2020/04/23 14:24:12 Account hostmaster@example.com is not registered. Use 'run' to register a new account.
2020/04/23 14:24:12 [INFO] acme: Registering account for hostmaster@example.com
2020/04/23 14:24:13 Could not complete registration
        acme: error: 429 :: POST :: https://acme-v02.api.letsencrypt.org/acme/new-acct :: urn:ietf:params:acme:error:rateLimited :: Error creating new account :: too many registrations for this IP: see https://letsencrypt.org/docs/rate-limits/, url:
```

**Comment by arianvp** [comment]
Could you share the logs of multiple individual cert services that are being renewed?


I have a hypothesis that if all are started at the same time; then there might actually be a race condition them all trying to create the account during startup. Once the account is created this shouldn't happen anymore though.

**Comment by rkoe** [comment]
Ok, I think I've found out:
- I'm using `hostmaster@mydomain` as `acme.certs.<name>.email` for the certificates, with `mydomain` replaced by the actual domain.
- This worked well with simp_le, but lego seems to create a new account for each email.
- Since I'm using 12 different email-adresses, it failed on the last two due to the max-10-accounts-per-3h-and-ip rate limit.

So:
- Will this happen at each certificate renewal, or only once during the simp_le-lego-transition?
- If this only happens during transition: Since I guess that this will not be fixed: At least a note/warning about this should be added to
  - the release notes for 20.03
  - the `acme.certs.<name>.email`-description

**Comment by arianvp** [comment]
Aha; yes I see. Thanks for the digging. This makes sense.

How did it work in `simp_le` @m1cr0man . Would one account have multiple emails associated to it?



> * Will this happen at each certificate renewal, or only once during the simp_le-lego-transition?

Only once. per account

> Adding a warning in release notes and to the description

This sounds like a good idea. Especially the remark that each `email` creates a new account...

**Comment by m1cr0man** [comment]
@arianvp I know you edited your message but I just wanted to confirm that yeah, it used to be one account per _certificate_, or in other words it was 1:1 between `security.acme.certs.<name>` and accounts. Now it is 1:1 with email addresses as observed above.

It's also worth pointing out that the `security.acme.server` is taken into consideration by lego too, so changing that will cause new accounts to be created too. This comes into play if you are swapping between staging and prod LE servers.

**Comment by arianvp** [comment]
So I think the conclusion is we're in a better shape than before.  But adding a warning about rate limits especially if you have more than 10 accounts could be added to the release notes so people do not run into this. 

I'll make a PR. @rkoe sounds good?

**Comment by rkoe** [comment]
Yes, thanks.

Additionally, a warning at the `acme.certs.<name>.email`-description [1] would be quite useful, since this would help users not to trigger the rate limit.

[1] e.g.: For each distinctive email-address, a separate account is registered at the ACME-server (e.g. let's encrypt). Note that ACME-servers usually have account rate limits (e.g. max. 10 per IP per 3 hours), so adding several acme-entries with different email-addresses may trigger those limits.

**Comment by m1cr0man** [comment]
This issue was resolved in #91121 and I did add a note about using different emails for lots of certs too.
